testdata/src/whitespace_no_multiline/no_multiline.go

```diff
@@ -48,3 +48,49 @@ func fn2(
 	fmt.Println("Hello, World")
 }
 
+// MultiFunc (FuncDecl) that's not `gofmt`:ed.
+func fn3(
+	arg1 int,
+	arg2 int,
+) { // want "unnecessary leading newline"
+  
+          
+  
+    // A comment.
+	fmt.Println("Hello, World")
+
+	if true { // want "unnecessary leading newline"
+
+
+
+        fmt.Println("No cmoments")
+
+
+    } // want "unnecessary trailing newline"
+       // Also at end
+
+
+  
+} // want "unnecessary trailing newline"
+
+// Regular func (FuncDecl) that's not `gofmt`:ed.
+func fn4() { // want "unnecessary leading newline"
+  
+          
+	  
+
+	fmt.Println("Hello, World")
+
+	if true { // want "unnecessary leading newline"
+
+
+
+        fmt.Println("No cmoments")
+
+
+    } // want "unnecessary trailing newline"
+
+
+
+        
+} // want "unnecessary trailing newline"
```

Comment:
I've added 2 tests 

```golang
// Regular func (FuncDecl) that's not `gofmt`:ed. with comments
func fn5() { 
 	// A comment that should still exist after this
	// This one should also still exist 
		  
	  

	fmt.Println("Hello, World")

	if true { 
		// A comment that should still exist after this
		// This one should also still exist



		fmt.Println("No cmoments")


	} // want "unnecessary trailing newline"



		
} // want "unnecessary trailing newline"

// Regular func (FuncDecl) that's not `gofmt`:ed. with comment blocks
func fn6() { 
	/*
		Lorem ipsum dolor sit amet, consectetur adipiscing elit.
		Curabitur ornare dolor at nulla ultrices cursus.
		Mauris pharetra metus ac condimentum sodales.
		Fusce viverra libero vitae tellus dictum, sed congue risus sodales.
	*/
		  
	  

	fmt.Println("Hello, World")

	if true {
		/* 
			Lorem ipsum dolor sit amet, consectetur adipiscing elit.
			Curabitur ornare dolor at nulla ultrices cursus.
			Mauris pharetra metus ac condimentum sodales.
			Fusce viverra libero vitae tellus dictum, sed congue risus sodales.
		*/ 


		fmt.Println("No cmoments")


	} // want "unnecessary trailing newline"



		
} // want "unnecessary trailing newline"
```

The thing is that the file is correctly formatted after running Whitespace but I am missing diagnostic warnings?

When I ran Whitespace, I go this output: (omitted out of context diagnostics)

``` 
/home/robin/dev/github.com/ultraware/whitespace/testdata/src/whitespace_no_multiline/no_multiline.go:121:1: unnecessary trailing newline
/home/robin/dev/github.com/ultraware/whitespace/testdata/src/whitespace_no_multiline/no_multiline.go:116:2: unnecessary trailing newline
/home/robin/dev/github.com/ultraware/whitespace/testdata/src/whitespace_no_multiline/no_multiline.go:153:1: unnecessary trailing newline
/home/robin/dev/github.com/ultraware/whitespace/testdata/src/whitespace_no_multiline/no_multiline.go:148:2: unnecessary trailing newline
```

This does not match the trend set by previous tests. For example [fn4](https://github.com/ultraware/whitespace/pull/8/commits/d9dd7ed2b510ff9409a7b39a40a9a98cca917160#diff-df158a7786b3b6e51a5ed562f2d8df485305e3ecf5db73bf0d5831ee7beb40a3R84) shows a proper diagnostic warning

```
/home/robin/dev/github.com/ultraware/whitespace/testdata/src/whitespace_no_multiline/no_multiline.go:84:49: unnecessary leading newline
```


Please correct me if I'm wrong there :sweat:

Reply:
These changes shouldn't really be touched by `whitespace` since they're not violating `whitespace` rules. A comment is seen as any statement and there are no whitespaces between the opening block and the first comment. However since the file isn't properly `gofmt`:ed the fixer will format it and replace all occurrences with more than one empty line with only one empty line.

I added your tests with proper `want` directives (just like you did) and the test passes and like you said formats the file correctly.

Regarding `fn4` (and I guess all test cases); I interpreted it as by design to _not_ see comments on the same line as a statement since that will still result in an empty line to the eye. This means that this is OK:

```go
func fn1() {
    // Comment is on new line

    fmt.Println("ok")
}
```

But this is not:

```go
func fn2() { // Comment is on same line as `LBrace`

    fmt.Println("ok")
}
```

Am I missing something here regarding either the intended way `whitespace` should work or your question? Do you think the tests I added in my last commit is what you expected?

Reply:
Ah, I see where I went wrong. I thought golangci was responsible for fixing. But it's analysis. Thanks for clearing that up

> This is interesting though, what do you think about this case?

That's a great question. First off, I have never seen that kind of comment in the wild

I assume the linter triggers in this case.

Personally I think this should not trigger a warning. But I'll have to gather some thoughs since this is undefined. I'll ask my colleagues asap.

What's your opinion @bombsimon

Reply:
> I thought golangci was responsible for fixing. But it's analysis. Thanks for clearing that up

Well, you're not wrong. Currently `golangci-lint` can't interpret the reports so it's both. When we run `whitespace` as a standalone linter with `-fix` it's the analysis package, when we run `golangci-lint` with `--fix` it's some custom code. That's the reason why I'm keeping which lines. But I don't think that's related, bottom line is that if it's one or more empty lines between the bracket line and the first comment or statement, both will report (and fix). If not, no matter how many lines are _after_ that initial comment or statement `whitespace` won't add a diagnostic.

> I assume the linter triggers in this case.

Yes, currently it does. But I think it's just a side effect and not intentional. In fact, I think the current check for comments is wrong, this line should check if the comment _starts_ on the same line, not if it ends.

I think I agree. Since this is OK:

```go
func Fn() { // Never use Fn with flux capacitor
    // Lorem ipsum dolor sit amet

    fmt.Println("")
}
```

I don't see why this should not be:

```go
func Fn() { /* Never use Fn with flux capacitor
    Lorem ipsum dolor sit amet */

    fmt.Println("")
}
```